### PR TITLE
Enable lazy objects mode in test runner

### PR DESCRIPTION
**Pull request by yinghuitan**
Release Note: enable lazy objects mode in test-runner.

Note: I will try @hermanventer's suggestion to move all lazy objects creation to prelude and hydrate them after that in future PR for a better way to deal with lazy objects testing issue.

1. Enables lazy objects mode in test runner. 
2. For JSON.stringify() and for...in constructs that hydrate lazy objects in global code, I added "force hydrate lazy objects" attribute in the generated code telling test-runner to hydrate the objects immediately
3. Disable lazy objects mode for 3 factorify tests and 4 abstract tests.
4. Skip lazy objects mode for react compiler, additional function tests.
5. Fix a bug which only shows up in lazy objects mode: a) if the target body is non-generator body and delay reason is generator body there is no need to wait for it. b) we should use targetBody instead of this._body in emitter if delayReason is a Value.

### CODE REVIEW

File: scripts/test-runner.js [hunk 1]
```diff
@@ -103,21 +104,49 @@ function execExternal(externalSpec, code) {
   return String(output.trim());
 }
 
+function argumentCodeWithLazyObjectSupport(code) {
```
Comment (simonhj) [inline scripts/test-runner.js]: Name not clear, did you mean "augment"?

File: src/intrinsics/ecma262/JSON.js [hunk 2]
```diff
@@ -505,9 +505,16 @@ export default function(realm: Realm): ObjectValue {
...
-      });
+      let result = AbstractValue.createTemporalFromTemplate(
+        realm,
+        JSONStringify,
+        StringValue,
+        [clonedValue],
+        {
+          kind: "JSON.stringify(...)",
+        },
+        true
```
Comment (NTillmann) [inline src/intrinsics/ecma262/JSON.js]: What is this? If it's really a thing, can it instead be moved into the named optional args?

File: src/values/AbstractValue.js [hunk 3]
```diff
@@ -516,7 +516,8 @@ export default class AbstractValue extends Value {
     template: PreludeGenerator => ({}) => BabelNodeExpression,
     resultType: typeof Value,
     operands: Array<Value>,
-    optionalArgs?: {| kind?: string, isPure?: boolean, skipInvariant?: boolean |}
+    optionalArgs?: {| kind?: string, isPure?: boolean, skipInvariant?: boolean |},
+    forceHydrateLazyObjects: boolean = false
```
Comment (NTillmann) [inline src/values/AbstractValue.js]: I don't fully understand the semantics of `forceHydrateLazyObjects` (I can see how it's all connected, but I don't have a good mental modal for it). Can I consider it a hack in place until we switch to @hermanventer's suggestion?

File: src/realm.js [hunk 4]
```diff
@@ -153,6 +153,7 @@ export class Realm {
     this.compatibility = opts.compatibility || "browser";
     this.maxStackDepth = opts.maxStackDepth || 225;
     this.omitInvariants = !!opts.omitInvariants;
+    this.lazyObjectsRuntime = opts.lazyObjectsRuntime;
```
Comment (NTillmann) [inline src/realm.js]: It's a bit weird that `lazyObjectsRuntime` is part of `RealmOptions`. Conceptually, it should really just concern the serializer. I understand that this is currently here to emit a magic comment when generating code (for loops) to enable testing.

File: scripts/test-runner.js (hunk 1 above)
Comment (yinghuitan) [reply scripts/test-runner.js]: Thank you! That's the word I am looking for.

File: src/realm.js (hunk 4 above)
Comment (yinghuitan) [reply src/realm.js]: Yes. It's for using from interpreter to emit that magic comment. I will add a comment.

File: src/values/AbstractValue.js (hunk 3 above)
Comment (yinghuitan) [reply src/values/AbstractValue.js]: Yes, it is a hack flag telling derive() method to emit that magic comment or not. If we are able to switch to @hermanventer's suggestion we should be able to remove this usage.

**Pull request closed without merging**